Subject: Allow one kernel binary with static 8250 and pmac_zilog
From: olh@suse.de
Patch-mainline: never

Allow co-existance of 8250=y and pmac_zilog=y.
Both try to claim ttyS0. 8250 comes first and zilog init fails.
The right fix would be a rewrite of the tty devnode handling with dynamic
minor assignment.
http://bugzilla.kernel.org/show_bug.cgi?id=1529

Signed-off-by: Olaf Hering <olh@suse.de>

 arch/powerpc/platforms/powermac/pci.c |   13 +++++++++++++
 drivers/serial/8250.c                 |    9 +++++++++
 drivers/serial/8250_pci.c             |    7 +++++++
 include/asm-powerpc/io.h              |    4 ++++
 include/asm-ppc/io.h                  |    5 +++++
 5 files changed, 38 insertions(+)

--- linux-2.6.19.orig/arch/powerpc/platforms/powermac/pci.c
+++ linux-2.6.19/arch/powerpc/platforms/powermac/pci.c
@@ -985,10 +985,23 @@ static int __init add_bridge(struct devi
 	return 0;
 }
 
+/*
+ * if both serial drivers exists, 8250 will be inited first
+ * it claims ttyS0 and pmac_zilog init fails
+ */
+#ifdef CONFIG_SERIAL_8250
+int do_not_probe_pc_legacy_8250;
+EXPORT_SYMBOL(do_not_probe_pc_legacy_8250);
+#endif
+
 void __init pmac_pcibios_fixup(void)
 {
 	struct pci_dev* dev = NULL;
 
+#if defined(CONFIG_SERIAL_8250) && defined(CONFIG_SERIAL_PMACZILOG)
+	do_not_probe_pc_legacy_8250 = 1;
+#endif
+
 	for_each_pci_dev(dev) {
 		/* Read interrupt from the device-tree */
 		pci_read_irq_line(dev);
--- linux-2.6.19.orig/drivers/serial/8250.c
+++ linux-2.6.19/drivers/serial/8250.c
@@ -46,6 +46,10 @@
 
 #include "8250.h"
 
+#ifndef NO_PC_LEGACY_SERIAL_8250
+#define do_not_probe_pc_legacy_8250 (0)
+#endif
+
 /*
  * Configuration:
  *   share_irqs - whether we pass IRQF_SHARED to request_irq().  This option
@@ -2334,6 +2338,8 @@ static struct console serial8250_console
 
 static int __init serial8250_console_init(void)
 {
+	if (do_not_probe_pc_legacy_8250)
+		return -ENODEV;
 	serial8250_isa_init_ports();
 	register_console(&serial8250_console);
 	return 0;
@@ -2641,6 +2647,9 @@ static int __init serial8250_init(void)
 {
 	int ret, i;
 
+	if (do_not_probe_pc_legacy_8250)
+		return -ENODEV;
+
 	if (nr_uarts > UART_NR)
 		nr_uarts = UART_NR;
 
--- linux-2.6.19.orig/include/asm-powerpc/io.h
+++ linux-2.6.19/include/asm-powerpc/io.h
@@ -413,6 +413,10 @@ static inline void __out_be64(volatile u
 #define dma_cache_wback(_start,_size)		do { } while (0)
 #define dma_cache_wback_inv(_start,_size)	do { } while (0)
 
+#if defined(CONFIG_PPC_PMAC) && defined(CONFIG_SERIAL_8250)
+#define NO_PC_LEGACY_SERIAL_8250 1
+extern int do_not_probe_pc_legacy_8250;
+#endif
 
 /*
  * Convert a physical pointer to a virtual kernel pointer for /dev/mem
--- linux-2.6.19.orig/include/asm-ppc/io.h
+++ linux-2.6.19/include/asm-ppc/io.h
@@ -535,6 +535,11 @@ extern void pci_iounmap(struct pci_dev *
 
 #endif
 
+#if defined(CONFIG_PPC_PMAC) && defined(CONFIG_SERIAL_8250)
+#define NO_PC_LEGACY_SERIAL_8250 1
+extern int do_not_probe_pc_legacy_8250;
+#endif
+
 /*
  * Convert a physical pointer to a virtual kernel pointer for /dev/mem
  * access
--- linux-2.6.19.orig/drivers/serial/8250_pci.c
+++ linux-2.6.19/drivers/serial/8250_pci.c
@@ -33,6 +33,10 @@
 
 #undef SERIAL_DEBUG_PCI
 
+#ifndef NO_PC_LEGACY_SERIAL_8250
+#define do_not_probe_pc_legacy_8250 (0)
+#endif
+
 /*
  * init function returns:
  *  > 0 - number of ports
@@ -2386,6 +2390,9 @@ static struct pci_driver serial_pci_driv
 
 static int __init serial8250_pci_init(void)
 {
+	if (do_not_probe_pc_legacy_8250)
+		return -ENODEV;
+
 	return pci_register_driver(&serial_pci_driver);
 }
 
